Reorganized guide + added tutorials

borjasotomayor opened the pull request:
This pull request includes two new tutorials (Linux and Git, each divided into a "basics" and "advanced" portion) that will be added to the guide. There are a number of other updates that relate to the organization of the site, and to some additional content that is not yet ready for review. So, right now, we just need to focus on the following files:

- `tutorials/linux-basics.rst`
- `tutorials/linux-advanced.rst`
- `tutorials/git-basics.rst`
- `tutorials/git-advanced.rst`

Note: these still include a few TODO blocks that can be skipped for now.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+Since one of the goals of this tutorial is for you to be able to use
+the CS department's Linux environment, we strongly suggest you work
+through this tutorial on a :ref:`UChicago CS software environment <software-environment>`
+(follow the link for more details on how to access such an environment).
```
This parenthetical feels unnecessary.

borjasotomayor replied:
I've removed it, but also tweaked the text that appears afterwards.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+
+If you are using a desktop environment (e.g., you are sitting in CSIL
+or you're using the Virtual Desktop), you can start a terminal by
+clicking on the Application icon (3x3 grid of dots) at the bottom left of
```
Since we are describing how the icon looks, I would suggest adding an image to supplement it.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+the CS department's Linux environment, we strongly suggest you work
+through this tutorial on a :ref:`UChicago CS software environment <software-environment>`
+(follow the link for more details on how to access such an environment).
+We particularly recommend using SSH to log into a CS Linux server from
```
Link the SSH article here.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+   :width: 650
+   :height: 250
+
+.. note::
```
Move note to the end of the subsection.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+called the prompt.  When you start typing, the characters you type
+will appear to the right of the ``$``.
+
+The program that runs within a terminal window and processes the
```
Make the distinction between the terminal and the shell more explicit.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
++-------------------+--------------------------------------------------------------+
+|``cd <path-name>`` |     change to the directory path-name                        |
++-------------------+--------------------------------------------------------------+
+|  ``cd ..``        |            move up/back one directory                        |
```
Maybe adding `cd -` could be useful? It's a big time saver when you have to return between two directories.

borjasotomayor replied:
Done, I've also added some text at the end of the section that describes `cd -` in more detail (with an example)

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+graphical system, we would double click on folders and occasionally
+click the "back" arrow. In order to change directories in
+the terminal, we use ``cd`` (change directory) followed by the name of
+the destination directory. (A note about notation: we will use text
```
How frequently does the guide clarify notation in its articles? To avoid having to repeat notes on notation across articles, perhaps a guide on notation could be written and referenced when necessary.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+
+To improve the readability of our examples, we will use ``$`` as the
+prompt rather than the full text ``username@computer:~$`` in the rest
+of this tutorial and, more generally, in the course going forward.  Keep in
```
Remove reference to a course

borjasotomayor replied:
Done

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+run the following commands::
+
+    cd
+    wget -nv https://uchicago-cs.github.io/dev-guide/_static/linux-tutorial-files.zip
```
The command returns a 404 error. I assume that is because this branch has not been deployed. It'd be worth checking that it works as expected once this version of the guide goes online.

borjasotomayor replied:
Yes, this should work once this branch is deployed. In the interim, the file can be found here: https://github.com/uchicago-cs/dev-guide/raw/tutorials/_static/linux-tutorial-files.zip

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+
+3. Make a new directory named ``backups`` using the ``mkdir`` command.
+
+4. Copy the file ``copy2.txt`` to the ``backups`` directory.
```
Did you mean `copy.txt`? We're asked to delete `copy2.txt` in 6.

borjasotomayor replied:
No, bear in mind that, after this step there will be a `copy2.txt` file in the `linux-tutorial-files` directory and in the `linux-tutorial-files/backups/` directory. Step 6 instructs the reader to remove the one in the `linux-tutorial-files` directory, under the rationale that "we've already made a backup of it".

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+Exercises
+^^^^^^^^^
+
+#. Navigate to the ``cmsc12100`` directory.  What do you see when you run ``ls tutorial*``?  What about ``ls tutorial*/*.py``?
```
Revise this exercise.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+files in your local directory). If it does not, use ``cd`` to navigate
+to the ``linux-tutorial-files`` directory.
+
+Note: somewhat counterintuitively, the menu bar for Visual Studio Code
```
Put this in a note block.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-basics.rst:
```diff
@@ -0,0 +1,618 @@
+Keyboard shortcuts
+~~~~~~~~~~~~~~~~~~
+
+Used at the Linux prompt, the keyboard shortcut ``Ctrl-P`` will roll
```
Change 'Linux prompt' to terminal. It might be useful to include that arrows can achieve the same as `Ctrl-P`/`Ctrl-N`.

borjasotomayor replied:
Done. There is some later text in the section that mentions that you can also use the arrow keys.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+if there is a group of commands that you want to execute sequentially.
+
+
+.. note::
```
It might be a good idea to specify that the semicolon is the line delimiter and that whitespace does not matter.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+The code in ``my_echo.py`` simply reads information from ``stdin`` and
+writes it back out to ``stdout``.  The redirection operators change
+the bindings of these streams from the keyboard and/or screen to files.
+We'll discuss ``stderr`` later in the term.
```
Reference to a course.

borjasotomayor replied:
Fixed.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+#. Run ``my_echo.py``  redirecting the input from ``test.txt`` and the output to ``output2.txt``.  Check the contents of ``output2.txt``.
+#. When you are done, remove ``output.txt`` and ``output2.txt``.
+
+By the way, if you run ``python3 my_echo.py`` without redirecting the input, it will patiently wait for you to type some input for it to echo.  Once you type some input and hit return, the program will echo your input, and then resume waiting for input.  It will continue to do so until you exit by typing ``Ctrl-d``.  Give it a try!
```
Place in note block

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+
+As we can see, neither of these options is particularly helpful. However, with
+piping, we can combine ``man -k`` with a powerful command line utility called
+``grep`` (see man pages) to find what we need::
```
Maybe add a link to the man page subsection of the basic linux tutorial?

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+
+Sometimes we want to restrict who can access certain resources on the file system.
+
+Most file systems assign 'File Permissions' (or just permissions) to specific users and groups of users. Unix is no different. File permissions dictate who can read (view), write (create/edit), and execute (run) files on a file system.
```
This is a nitpick, but I noticed that the guide has inconsistent formatting for vocabulary terms. Sometimes quotes are used and sometimes italics are used. Again, this is a minor point and it should not take priority over the actual content of the guide, but if the intention is for this website to function as a centralized resource we should consider consistent formatting. Because I can imagine this being a project maintained by other people in the future, establishing a style guide for articles could be useful. The [Wikipedia MOS](https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style) is a good example of what I mean.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+
+
+First thing we can notice is that the owner of the file is a user
+named ``root``. (FYI, ``root`` is a name for an account that has access
```
Put the parenthetical in a note block.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+Exercise
+~~~~~~~~
+
+#. Verify this claim by running ``ls -l backups/copy2.txt`` and ``ls -ld  backups`` in your ``linux-tutorial-files`` directory.
```
This assumes that the student has done the basic linux exercises. Because the tutorial has been split into two parts, I do not think this is a safe assumption to make.

borjasotomayor replied:
I've added a note in the TODO block at the top to remind myself to add some text saying that they need to download the tutorial files (but that they don't need to work through the Linux Basics tutorial to work through this one). I've also added a note in the exercise, providing the exact commands to create the necessary directory and file, in case someone didn't work through the Linux Basics tutorial.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+
+    chgrp <new group> <path to file>
+
+It is unlikely that you will need to use these two commands for
```
Reference to a course

borjasotomayor replied:
Fixed.

victor-almaraz commented on tutorials/linux-advanced.rst:
```diff
@@ -0,0 +1,344 @@
+~~~~~~~~~
+
+#. Run ``echo "Hello!" > testfile`` to construct ``testfile``.  Look at the permissions using ``ls -l``.
+#. Change the permissions on ``testfile`` to allow and read access for others.  Run ``ls -l testfile`` to check the new permissions.
```
Revise this one. I think there is something missing before the and.

borjasotomayor replied:
Fixed (it should say "write and read")

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+    then clicking on "Your profile" or "Your repositories".
+
+
+Setting up SSH access
```
I would suggest putting this first. It is a bit awkward starting to set up a repo only to take a detour to set up the SSH key.

borjasotomayor replied:
Done, I've reordered the sections so the SSH section takes place before creating a repository.

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+Copy the whole output to the clipboard; you can do this by clicking and dragging the mouse from the first
+character to the last character, and then pressing Ctrl-Shift-C.
+
+Then, paste the key into the "Key" field on the GitHub page. Then click on the green "Add SSH Key"
```
It might be useful to provide an image of what the page looks like.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+   there is a URL field with two buttons: HTTPS and SSH.
+   Make sure that “SSH” is selected.
+
+Now, from inside your tutorial directory, run the commands that appear
```
Include a code box that shows what the commands should look like, at least at the time of writing.

borjasotomayor replied:
I don't think we want to do this, as GitHub could update those instructions in the future, and it's important that they run what appears on GitHub. That said, I've included a screenshot of what a new repository looks like, so they have a sense that they're in the right page.

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+to your repository’s page on GitHub, you should now see it contains a
+``README.md`` file. If you click on it, you can see its contents.
+
+Creating a commit
```
I would suggest making a more explicit distinction between staging and committing. Particularly how you can add and remove things from staging, and how all staged changes are made once a commit is made.

victor-almaraz replied:
In other words, I think it would be useful to explain a bit more of the ideas behind git rather than just stating what commands do what.

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+
+.. note::
+
+   In older versions of Git, ``git status`` may refer to the ``git checkout`` command.
```
Did you meant `git restore` instead of `git status`?

borjasotomayor replied:
No, the note refers to the output from `git status` (which, in older versions, will tell you to run `git checkout` instead of `git restore`)

victor-almaraz commented on tutorials/git-basics.rst:
```diff
@@ -0,0 +1,909 @@
+
+   git restore --staged test.txt
+
+.. note::
```
Did you meant `git restore` instead of `git status`?

borjasotomayor replied:
Same as previous comment; the note refers to the output from `git status`

victor-almaraz commented on tutorials/git-advanced.rst:
```diff
@@ -0,0 +1,604 @@
+separate functions, ``bar()`` and ``baz()``).
+
+To experiment with branches, start by adding a file called ``echo.c``
+to your repository, with the following contents:
```
You should clarify if this is the repo from the first half of the tutorial.

borjasotomayor replied:
I've added a note in the TODO at the top saying that we'll need to tell them to create an empty repository for this tutorial (if they haven't worked through the Git Basics tutorial)

victor-almaraz commented on tutorials/git-advanced.rst:
```diff
@@ -0,0 +1,604 @@
+    Branch 'add-author' set up to track remote branch 'add-author' from 'origin'.
+
+Note: You should ignore the "pull request" instructions in the above message.
+We will cover pull requests later in the quarter.
```
Reference to a course. Adding a section on pull requests would be a good idea.

borjasotomayor replied:
Fixed the reference to the course. Please add an issue related to adding a section on pull requests. I agree it makes sense to add one, but it will be a bit of work to do so.

victor-almaraz commented on tutorials/git-advanced.rst:
```diff
@@ -0,0 +1,604 @@
+        return 1;
+    }
+
+In general, any time you see something like this:
```
It would be a good idea to include a mention of how modern IDEs give you the option to select which changes to keep with a button as opposed to manually having to update the file.

borjasotomayor replied:
Done.

victor-almaraz commented on tutorials/git-advanced.rst:
```diff
@@ -0,0 +1,604 @@
+
+        Updated buffer size to 1000
+
+
```
Having taken CS 220 I learned about CI and it made my life a lot easier. I would suggest adding a section not necessarily on CI per se, but on the ideas behind it. Telling students about branches doesn't mean they'll know how to use them effectively. I think it would be good to tell them to use branches in conjunction with frequent and isolated changes.

borjasotomayor replied:
Please create an issue related to adding a section on CI. Like the section on pull requests, I think this is a good idea, but will take a bit more time to do.

borjasotomayor commented:
I've pushed the newest version of the VSCode+SSH guide (`environment/vscode-ssh.rst`), so that is now ready for review too. 

@vaargueta: You don't need to run through all the SSH installation instructions (in Windows, Mac, and Linux), but you may want to read through them to make sure they're clear, etc. By the way, I suspect we may end up moving a lot of the SSH installation instructions into the "Remote SSH Access" page.

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+Windows 10
+~~~~~~~~~~
+
+*These instructions are specific to Windows 10. If you are running Windows 7 or Windows 8, please contact us on Piazza for instructions.*
```
Revise reference to Piazza.

borjasotomayor replied:
Fixed in branch `vaa/tutorial`

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+
+.. figure:: code-img/install-ssh-win10-10.png
+
+You should be prompted for your password. If you are not, check that you followed the SSH installation steps correctly, and try again. If you are still not prompted for your password, ask us about it on Piazza.
```
Reference to Piazza.

borjasotomayor replied:
Fixed in branch `vaa/tutorial`

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+
+*These instructions are specific to Windows 10. If you are running Windows 7 or Windows 8, please contact us on Piazza for instructions.*
+
+In this step, you will install Windows OpenSSH Client.
```
I don't think I ever had to do this. I've always used ssh on GitBash and that seemed to work fine. Someone should check if that is the case because if so the set up is far easier.

borjasotomayor replied:
Fixed in branch `vaa/tutorial`

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+
+**Initial setup**
+
+You only need to follow the steps in this section once (or more accurately, once per computer that you will use to connect remotely). If you've already done this part, you can continue to "Connecting".
```
Parenthetical can be rewritten into sentence.

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+    username@computer:~$
+
+
+Editing text files
```
I'll have to double check this, but I think VS Code lets you use the file explorer side panel even through SSH. If that is the case, it would be worth mentioning.

victor-almaraz commented on environment/vscode-ssh.rst:
```diff
@@ -0,0 +1,410 @@
+
+where ``username`` is replaced by your CNetID and ``LINUX_SERVER`` is replaced with your assigned linux server. You should be prompted for the password associated with your CNetID. Then you should be able to run terminal commands on the campus Linux computers.
+
+Step 3: Install Extensions for VSCode
```
If this is meant to be a general use guide, perhaps we should remove this section on C/C++ extensions. Or perhaps move it to the end of the article as a general guide on adding extensions for all languages.

borjasotomayor replied:
Fixed in branch `vaa/tutorial` (I generalized it so it doesn't refer just to C/C++)

victor-almaraz commented:
@borjasotomayor I think moving the SSH installation instructions to the SSH article is a good idea.

borjasotomayor commented:
I have reviewed the changes/additions in the `vaa/tutorial` branch. They mostly look good to me, but I've used the opportunity to refactor the content into its own section on the guide (instead of having a single massive page covering installation, SSH setup, configuration, etc.) I've also added some of the missing content in the Linux/Git tutorials, and gone ahead and changed the title to UChicago CS Student Resource Guide.

At this point, the reorganized guide should be almost ready for an initial release.
